### DISCUSSION TITLE
How to utilize carry chains and MUX8 and dffesr when sythesis user design using yosys and PnR using nextpnr?

### DISCUSSION THREAD

**Discussion by keepzhen** [FPGA-Research/FABulous#260] [Q&A]
Hi,
I want to implement a complex user desgin that may benefit from the use of carry chain and MUX8(eg. to cascade mutiple LUT4 to LUT7).Also I know that less LUT4 will be used if DFF in FABULOUS_LC have enable port and set\reset port.

After learning the example(implement counter.v and sequential_16bits.v),it works well but it seems only can auto infer to LUT4 and LUT_FF(without enable or set).I notice that in actual FABulous FPGA fabric design,LUT4AB tile contains carry chain and DFF with E\SR and MUX8 .  

I have tried using command in yosys (yosys dfflegalize -cell \$_SDFFCE_PP?P_ 0 -cell \$_DLATCH_?_ x) to map into LUTFF_ESR/S succefully and generates *.josn file , but it fails in npnr and could not generate bitstream.I guess I maybe ignore some steps to sythesis and PnR.

Could you give some suggestion to how to utilize LUTFF_ESR/S,besides carry chain and MUX8 in LUT4AB tile ?

**Comment by KelvinChung2000** [comment]
You might want to implement your own techmap, primitive and a custom synthesis script. 

The techmap pass is where yosys inferring the LUT4 and LUT_FF. You can find all the related files in [here](https://github.com/YosysHQ/yosys/tree/main/techlibs/fabulous). 

What you might want to do is synthesise the design to uses LUT7, then use techmap to replace the LUT7 with your LUT4+CC as LUT7 custom primitives. 

There is the [guide](https://tomverbeure.github.io/2022/11/18/Primitive-Transformations-with-Yosys-Techmap.html) on what techmap does and how to implement your custom techmap. 

I am thinking there might be a problem during place and route, as the nextpnr do not know the 2 LUT4+CC need to be placed next to each other, and you might also need to provide a custom packaging script.

**Comment by dirk-koch** [comment]
The present FABulous CLBs have a BEL MUX8LUT_frame_config that implements a F5, F6, F7 Mux, similar to Xilinx Virtex-II.
We haven't added support in Yosys for that, but the figure will help you.
The two select bits select between 4xF5Mux, 2xF6Mux or 1xF7Mux. Of course unused (e.g., F5Muxes) can be left unconnected. 
The FABulous routing graph is designed to support that nicely.
![MUX8LUT_frame_config](https://github.com/user-attachments/assets/453dfb93-743d-4ab1-b88b-9afcf3c936cf)
Let us know if you need help with this.
